### PR TITLE
GH-3437: Add `MessageRecoverer` to AMQP Inbounds

### LINKED ISSUE #3437: Add MessageRecoverer into AMQP inbound channel adapters
**Expected Behavior**

The `AmqpInboundGateway` and `AmqpInboundChannelAdapter` provide a `MessageRecoverer` option to inject for more convenient way to deal with these channel adapters at the configuration phase.
Same option must be exposed via Java DSL.

**Current Behavior**

There is only a `RecoveryCallback` option at the moment alongside with the `RetryTemplate` which is not so useful in context of failed AMQP message to process.

It is possible though to inject a `RetryInterceptor` via  Spring AMQP `RetryInterceptorBuilder` API and `MessageRecoverer` strategy into the listener container for the mentioned AMQP inbound channel adapters.

**Context**

See more info on Gitter chat: https://gitter.im/spring-projects/spring-integration?at=5fc71bb5924a486dd9e0c94e

### PR DESCRIPTION
**Pull request by artembilan**
Fixes https://github.com/spring-projects/spring-integration/issues/3437

For better end-user experience with AMQP Inbound Endpoints, expose
a `MessageRecoverer` option
* Test the feature and document this new option

### CODE REVIEW

File: src/reference/asciidoc/amqp.adoc [hunk 1]
```diff
@@ -199,6 +199,9 @@ The JMS inbound channel adapter is using a `JmsDestinationPollingSource` under t
 The AMQP inbound channel adapter uses an `AbstractMessageListenerContainer` and is message driven.
 In that regard, it is more similar to the JMS message-driven channel adapter.
 
+Starting with version 5.5, the `AmqpInboundChannelAdapter` can configured with a `org.springframework.amqp.rabbit.retry.MessageRecoverer` strategy which is used in the `RecoveryCallback` when retry operation is called internally.
```
Comment (garyrussell) [inline src/reference/asciidoc/amqp.adoc:202]:
```suggestion
Starting with version 5.5, the `AmqpInboundChannelAdapter` can be configured with an `org.springframework.amqp.rabbit.retry.MessageRecoverer` strategy which is used in the `RecoveryCallback` when the retry operation is called internally.
```

File: src/reference/asciidoc/amqp.adoc [hunk 2]
```diff
@@ -395,6 +400,9 @@ if you anticipate cases when no `replyTo` property exists in the request message
 
 See the note in <<amqp-inbound-channel-adapter>> about configuring the `listener-container` attribute.
 
+Starting with version 5.5, the `AmqpInboundChannelAdapter` can configured with an `org.springframework.amqp.rabbit.retry.MessageRecoverer` strategy which is used in the `RecoveryCallback` when retry operation is called internally.
```
Comment (garyrussell) [inline src/reference/asciidoc/amqp.adoc:403]: Same change as above.

File: src/reference/asciidoc/whats-new.adoc [hunk 3]
```diff
@@ -17,3 +17,10 @@ If you are interested in more details, see the Issue Tracker tickets that were r
 
 [[x5.5-general]]
 === General Changes
+
+[[x5.5-amqp]]
+==== AMQP Changes
+
+The `AmqpInboundChannelAdapter` and `AmqpInboundGateway` (and respective Java DSL builders) now support an `org.springframework.amqp.rabbit.retry.MessageRecoverer` as an AMQP-specific alternative to a general purpose `RecoveryCallback`.
```
Comment (garyrussell) [inline src/reference/asciidoc/whats-new.adoc:24]:
```suggestion
The `AmqpInboundChannelAdapter` and `AmqpInboundGateway` (and the respective Java DSL builders) now support an `org.springframework.amqp.rabbit.retry.MessageRecoverer` as an AMQP-specific alternative to the general purpose `RecoveryCallback`.
```

**Review by garyrussell — CHANGES_REQUESTED** [review]
Looks good; just needs a bit of doc polishing.

File: src/reference/asciidoc/amqp.adoc (hunk 2 above)
Comment (artembilan) [reply src/reference/asciidoc/amqp.adoc:403]:
```suggestion
Starting with version 5.5, the `AmqpInboundChannelAdapter` can be configured with an `org.springframework.amqp.rabbit.retry.MessageRecoverer` strategy which is used in the `RecoveryCallback` when the retry operation is called internally.
```

**Review by garyrussell — APPROVED** [review]

**Pull request merged**